File: AssignmentSubmissionManagement.API/Models/Assignment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AssignmentSubmissionManagement.API.Models
{
    public class Assignment
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [MaxLength(200)]
        public string Title { get; set; } = string.Empty;

        [Required]
        public string Description { get; set; } = string.Empty;

        [Required]
        public DateTime Deadline { get; set; }

        [Required]
        [Range(1, 1000)]
        public int MaximumMarks { get; set; }

        // Draft = false, Published = true
        public bool IsPublished { get; set; } = false;

        // Student deadline-এর আগে submission edit করতে পারবে কি না
        public bool AllowSubmissionUpdate { get; set; } = true;

        [Required]
        public int TeacherId { get; set; }

        [Required]
        public int ClassRoomId { get; set; }

        [Required]
        public int SubjectId { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime? UpdatedAt { get; set; }

        // Navigation Properties
        [ForeignKey(nameof(TeacherId))]
        public User? Teacher { get; set; }

        [ForeignKey(nameof(ClassRoomId))]
        public ClassRoom? ClassRoom { get; set; }

        [ForeignKey(nameof(SubjectId))]
        public Subject? Subject { get; set; }
    }
}

File: assignment-submission-frontend/src/types/assignment.ts
import api from "@/config/api";

export interface Assignment {
    id: number;
    title: string;
    description: string;
    deadline: string;
    maximumMarks: number;

    isPublished?: boolean;
    allowSubmissionUpdate: boolean;

    teacherId?: number;
    teacherName?: string;

    classRoomId?: number;
    classRoomName?: string;

    subjectId?: number;
    subjectName?: string;

    createdAt?: string;
    updatedAt?: string;
}

export interface CreateAssignmentRequest {
    title: string;
    description: string;
    deadline: string;
    maximumMarks: number;

    isPublished: boolean;
    allowSubmissionUpdate: boolean;

    classRoomId: number;
    subjectId: number;
}


// ========================================
// ADMIN - View all assignments
// GET /api/Assignments/all
// ========================================
export const getAllAssignments =
    async (): Promise<Assignment[]> => {

        const response =
            await api.get<Assignment[]>("/Assignments/all");

        return response.data;
    };


// ========================================
// TEACHER - View own assignments
// GET /api/Assignments/my
// ========================================
export const getMyAssignments =
    async (): Promise<Assignment[]> => {

        const response =
            await api.get<Assignment[]>("/Assignments/my");

        return response.data;
    };


// ========================================
// STUDENT - View published assignments
// GET /api/Assignments/student
// ========================================
export const getStudentAssignments =
    async (): Promise<Assignment[]> => {

        const response =
            await api.get<Assignment[]>("/Assignments/student");

        return response.data;
    };


// ========================================
// Role Based Assignment Loader
// ========================================
export const getAssignmentsByRole =
    async (role: string): Promise<Assignment[]> => {

        switch (role) {

            case "Admin":
                return await getAllAssignments();

            case "Teacher":
                return await getMyAssignments();

            case "Student":
                return await getStudentAssignments();

            default:
                return [];
        }
    };


// ========================================
// TEACHER - Create Assignment
// POST /api/Assignments
// ========================================
export const createAssignment =
    async (
        data: CreateAssignmentRequest
    ) => {

        const response =
            await api.post(
                "/Assignments",
                data
            );

        return response.data;
    };